src/features/device-state/routes/fleet-state-get-v3.ts

```diff
@@ -0,0 +1,372 @@
+		(fleetService: AnyObject) => fleetService.id === serviceId,
+	);
+};
+
+function buildFleetAppFromRelease(
+	fleet: AnyObject,
+	release: AnyObject,
+	config: Dictionary<string>,
+): NonNullable<LocalStateApp['releases']> {
+	let composition: AnyObject = {};
+	const services: NonNullable<LocalStateApp['releases']>[string]['services'] =
+		{};
+
+	// Parse the composition to forward values to the service entries
+	if (_.isObject(release.composition)) {
+		composition = release.composition;
+	} else {
+		try {
+			composition = JSON.parse(release.composition);
+		} catch {
+			composition = {};
+		}
+	}
+	for (const ipr of release.contains__image as AnyObject[]) {
+		// extract the per-image information
+		const image = ipr.image[0];
+		const svc = serviceBuildByImage(fleet, image);
+		if (svc == null) {
+			throw new Error(
+				`Could not find service build for fleet: '${fleet.uuid}', image: '${
+					image?.id
+				}', service: '${JSON.stringify(image?.is_a_build_of__service)}'`,
+			);
+		}
+
+		const environment: Dictionary<string> = {};
+		varListInsert(ipr.image_environment_variable, environment);
+		varListInsert(fleet.application_environment_variable, environment);
+		varListInsert(svc.application__has__service_name__has__name, environment);
+
+		const labels: Dictionary<string> = {};
+		for (const { label_name, value } of [
+			...ipr.image_label,
+			...svc.application__has__service_name__has__label_name,
+		] as Array<{ label_name: string; value: string }>) {
+			labels[label_name] = value;
+		}
+
+		_.each(ConfigurationVarsToLabels, (labelName, confName) => {
+			if (confName in config && !(labelName in labels)) {
+				labels[labelName] = config[confName];
+			}
+		});
+
+		const imgRegistry =
+			image.is_stored_at__image_location +
+			(image.content_hash != null ? `@${image.content_hash}` : '');
+
+		services[svc.service_name] = {
+			id: svc.id,
+			image_id: image.id,
+			image: formatImageLocation(imgRegistry),
+			environment,
+			labels,
+		};
+		// Don't send a null contract as this is a waste
+		// of bandwidth (a null contract is the same as
+		// the lack of a contract field)
+		if (image.contract != null) {
+			services[svc.service_name].contract = image.contract;
+		}
+
+		if (composition?.services?.[svc.service_name] != null) {
+			const compositionService = composition.services[svc.service_name];
+			// We remove the `build` properly explicitly as it's expected to be present
+			// for the builder, but makes no sense for the supervisor to support
+			delete compositionService.build;
+			services[svc.service_name].composition = compositionService;
+		}
+	}
+
+	return {
+		[release.commit]: {
+			id: release.id,
+			...(Object.keys(services).length > 0 ? { services } : undefined),
+			...(composition.networks != null && { networks: composition.networks }),
+			...(composition.volumes != null && { volumes: composition.volumes }),
+		},
+	};
+}
+
+const releaseExpand = {
```

Comment:
Can this be shared with device-state-v3?

Reply:
yes, thx!